custom_components/tesla_custom/base.py

```diff
@@ -0,0 +1,328 @@
+    ) -> None:
+        """Initialise the Tesla car device."""
+        super().__init__(hass, coordinator)
+        self.car = TeslaCar(car, coordinator)
+
+    async def update_controller(
+        self, *, wake_if_asleep: bool = False, force: bool = True, blocking: bool = True
+    ):
+        """Get the latest data from Tesla.
+
+        This does a controller update,
+        then a coordinator update.
+        the coordinator triggers a call to the refresh function.
+
+        Setting the Blocking param to False will create a background task for the update.
+        """
+
+        if blocking is False:
+            await self.hass.async_create_task(
+                self.update_controller(wake_if_asleep=wake_if_asleep, force=force)
+            )
+            return
+
+        await self._coordinator.controller.update(
+            self.car.id, wake_if_asleep=wake_if_asleep, force=force
+        )
+        await self._coordinator.async_refresh()
+
+    @property
+    def available(self) -> str:
+        """Return the Availability of Data."""
+        return self.car.state != {}
+
+    @property
+    def name(self) -> str:
+        """Return the name of car device."""
+        if self._name is None:
+            self._name = (
+                f"{self.car.display_name} {self.type}"
+                if self.car.display_name is not None
+                and self.car.display_name != self.car.vin[-6:]
+                else f"Tesla Model {str(self.car.vin[3]).upper()} {self.type}"
+            )
+
+        return self._name
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID for car device."""
+        if self._unique_id is None:
+            self._unique_id = slugify(
+                f"Tesla Model {str(self.car.vin[3]).upper()} {self.car.vin[-6:]} {self.type}"
+            )
+
+        return self._unique_id
+
+    @property
+    def device_info(self):
+        """Return the device_info of the device."""
+        return {
+            "identifiers": {(DOMAIN, self.car.id)},
+            "name": self.car.name,
+            "manufacturer": "Tesla",
+            "model": self.car.type,
+            "sw_version": self.car.version,
+        }
+
+    @property
+    def assumed_state(self) -> bool:
+        # pylint: disable=protected-access
+        """Return whether the data is from an online vehicle."""
+        return not self._coordinator.controller.is_car_online(vin=self.car.vin) and (
+            self._coordinator.controller.get_last_update_time(vin=self.car.vin)
+            - self._coordinator.controller.get_last_wake_up_time(vin=self.car.vin)
+            > self._coordinator.controller.update_interval
+        )
+
+
+class TeslaCar:
+    """Data Holder for all Car Data.
+
+    Exists simply so we don't have a bunch of attributes on the top level Entity.
+    """
+
+    def __init__(self, car: dict, coordinator: TeslaDataUpdateCoordinator) -> None:
```

Comment:
Not sure how much it matters but `car` is passed in as a `list` containing `dict`. I could be missing something but `energysites` is the same way since there can be more than one, the Tesla API sends a list.

Reply:
Disregard, I see now that it's actually a dict that gets passed in when the entities are initialized.